proposals/webidl-bindings/binary.bs

```diff
@@ -0,0 +1,843 @@
+=================
+
+A complete encoding of a webIDL consists of a sequence of top-level
+elements. These top-level elements can be viewed as defining a set of interfaces
+and/or types; overall, the set of top-level elements define a *scope* in the
+form of an array; elements of which are referenced by numeric index.
+
+
+```
+WebIDL => u32(C) TopLevel{C}
+
+TopLevel => TypeDef
+TopLevel => InterfaceDef
+TopLevel => DictionaryDef
+TopLevel => EnumerationDef
+TopLevel => NamespaceDef
+```
+
+Top-level Index {#top-level-index}
+-------------
+
+Each element of the top-level has an index -- which is the relative position of
+the element in the array of all top-level elements.
+
+TypeDef {#type-def}
+------
+
+A `TypeDef` is a way of giving a name to a type:
+
+```
+TypeDef => 0x01 String Type
+```
+
+<div class="example">
+For example, the webIDL type definition:
+<pre highlight="webidl">
+typedef sequence&lt;long> rows;
+</pre>
+is represented using
+
+```
+0x01 0x04 r o w s 0x69 0x7a
+```
+</div>
+
+
+InterfaceDef {#interface-def}
+--------
+
+An `InterfaceDef` gives a name to an interface.
+
+```
+InterfaceType => 0x02 String Type u32(C) InterfaceMember{C}
+```
+
+DictionaryDef {#dictionary-def}
+-------------
+
+```
+DictionaryDef => 0x03 String Type u32(F) DictionaryMember{F}
+
+DictionaryMember => WriteableDictionaryMember
+DictionaryMember => ReadOnlyDictionaryMember
+DictionaryMember => 0x5a String DictionaryMember
+
+WriteableDictionaryMember => 0x01 String Type
+ReadOnlyDictionaryMember => 0x02 String Type
+```
+
+Note: This specification does not support partially specified dictionaries.
+
+EnumerationDef {#enumeration-def}
+----------------
+
+An `EnumerationDef` defines a type by enumerating its values as one of a set of
+strings.
+
+```
+EnumerationDef => 0x04 String u32(C) String{C}
+```
+
+NamespaceDef {#namespace-def}
+-----------
+
+A `NameSpace` simultaneously defines an `Interface` and a singleton that
+implements that interface.
+
+```
+NameSpaceDef => 0x05 String u32(C) NameSpaceMember{C}
+```
+
+The `NameSpaceMember`s are a subset of the `InterfaceMember`s that refer to
+regular operations and attributes.
+
+```
+NameSpaceMember => 0x02 String Type
+NameSpaceMember => 0x5a String NameSpaceMember
+```
+
+Note: all elemements of a `NameSpace` are read-only in nature.
+
+Type {#type}
+=====
+
+The encoding of types follows a particular pattern: if the type is one
+that is part of webIDL itself then it is encoded using a small number
+that denotes a negative number in the LEB encoding. If the type is a
+reference to a user-defined type; i.e., its a reference to a type that
+is declared in one of InterfaceDef, DictionaryDef, EnumerationDef then
+it is represented as a positive number: the index of its definition in
+the top-level sequence.
+
+```
+Type => AnyType
+Type => NumericType
+Type => StringType
+Type => ObjectType
+Type => ErrorType
+Type => ParameterizedType
+Type => BufferSourceType
+Type => CompoundType
```

Comment:
We should be de-duplicating / referring to compound types by index ([like my experimental implementation does](https://github.com/rustwasm/wasm-webidl-bindings/blob/master/BINARY.md#references-to-web-idl-types)) to save on code size.

In fact, I can't find where `Type` transitively goes to some sort of reference to an already-defined type at all, so for example a `FunctionType` cannot take a previously defined dictionary type as a parameter under these current rules. Is some sort of a production like `Type => ReferenceToPreviouslyDefinedType` missing?